java/code/src/com/redhat/rhn/frontend/servlets/ajax/AjaxHandlerServlet.java

```diff
@@ -0,0 +1,189 @@
+        // (Admin -> Setup Wizard -> Organization Credentials)
+        HANDLERS.put("verify-mirror-credentials", (req, resp) -> {
+            VerifyMirrorCredentialsDto dto = parseBody(req, VerifyMirrorCredentialsDto.class);
+            return mirrorCredentialsRenderer.verifyCredentials(req, resp, dto.getId(), dto.isRefresh());
+        });
+        HANDLERS.put("list-mirror-subscriptions", (req, resp) -> {
+            ListMirrorSubscriptionsDto dto = parseBody(req, ListMirrorSubscriptionsDto.class);
+            return mirrorCredentialsRenderer.listSubscriptions(req, resp, dto.getSubscriptionsId());
+        });
+        HANDLERS.put("render-mirror-credentials", mirrorCredentialsRenderer::renderCredentials);
+        HANDLERS.put("save-mirror-credentials", (req, resp) -> {
+            SaveMirrorCredentialsDto dto = parseBody(req, SaveMirrorCredentialsDto.class);
+            return mirrorCredentialsRenderer.saveCredentials(req, dto.getId(), dto.getUser(), dto.getPassword());
+        });
+        HANDLERS.put("make-primary-mirror-credentials", (req, resp) -> {
+            MakePrimaryMirrorCredentialsDto dto = parseBody(req, MakePrimaryMirrorCredentialsDto.class);
+            return mirrorCredentialsRenderer.makePrimaryCredentials(req, resp, dto.getId());
+        });
+        HANDLERS.put("delete-mirror-credentials", (req, resp) -> {
+            DeleteMirrorCredentialsDto dto = parseBody(req, DeleteMirrorCredentialsDto.class);
+            return mirrorCredentialsRenderer.deleteCredentials(req, resp, dto.getId());
+        });
+
+        // The following 2 handlers are used in the Action Chains page (Schedule -> Action Chains)
+        HANDLERS.put("action-chain-entries", (req, resp) -> {
+            ActionChainEntriesDto dto = parseBody(req, ActionChainEntriesDto.class);
+            return actionChainEntryRenderer.renderAsync(req, resp, dto.getActionChainId(), dto.getSortOrder());
+        });
+        HANDLERS.put("action-chain-save", ((req, resp) -> {
+            ActionChainSaveDto dto = parseBody(req, ActionChainSaveDto.class);
+            return actionChainSaveAction.save(
+                dto.getActionChainId(),
+                dto.getLabel(),
+                dto.getDeletedEntries(),
+                dto.getDeletedSortOrders(),
+                dto.getReorderedSortOrders(),
+                req
+            );
+        }));
+
+        // The following handler is used in all the pages using the rl:selectablecolumn custom tag
+        HANDLERS.put("item-selector", (req, resp) -> {
+            ItemSelectorDto item = parseBody(req, ItemSelectorDto.class);
+            return itemSelector.select(req, item.getLabel(), item.getValues(), item.isChecked());
+        });
+    }
+
+    @Override
+    public void doPost(HttpServletRequest req, HttpServletResponse resp) {
+        String url = req.getRequestURI().split(AJAX_PREFIX)[1];
+        try {
+            Object result = HANDLERS.get(url).doProcess(req, resp);
```

Comment:
I'm not entirely familiar with the context of the discussion here, but as someone who consumes a lot of APIs, if I feed it a non-existent URL, I would expect to get back 404, not a NPE.

Reply:
I agree on converting it to another exception, with a clear message. Do it now would be simple and fast, but if some L3 comes with a null point in this part it will take much more time to debug it.